### DISCUSSION TITLE
[Help]: Center slides in case their total width is less than the container

### DISCUSSION THREAD

**Discussion by rudokotrla** [davidjerleke/embla-carousel#1068] [Help] [answered]
### Summary

Hello, I have a related (but not 100% same question) so I will dump it here instead of creating a new thread.
I would like to make slides centered but **only in case the total sum of slides' width (+ the gaps in between) is less than the available container viewport width**. The `align: 'center'` option itself doesn't work. 
So instead of this: 
<img width="671" alt="Screenshot 2024-11-22 at 09 04 17" src="https://github.com/user-attachments/assets/27bf9183-aa40-4d33-9f7f-206ce592ebfc">
I would like the carousel to looks like this:
<img width="656" alt="Screenshot 2024-11-22 at 09 04 35" src="https://github.com/user-attachments/assets/8db772cf-8e64-4eeb-a651-7f0a347eb43f">
What I've tried is setting `justify-content: center` on the container (with `flex` layout) but this breaks the behaviour when there is more items (so scrolling is needed):
<img width="529" alt="Screenshot 2024-11-22 at 09 14 42" src="https://github.com/user-attachments/assets/1de3814b-c53b-4c16-9e84-d2246d135436">

I've checked out the proposed solution to the original question of this thread but it makes the carousel slides centered in all cases so also in case when there is enough slides to scroll, which is not what I want. In that case I would like to have practically `align: 'start'` behaviour.
<img width="549" alt="Screenshot 2024-11-22 at 09 38 03" src="https://github.com/user-attachments/assets/41da4378-58ee-4094-bdbe-b399619f8e10">

So far I settled up with a bit hacky solution that sets the `justify-content: center` style on the container only when `api.slidesInVIew().length` equal to `api.slidesNodes().length` (all slides). It is done by adding event handler to the `slidesInView` that checks the condition and updates React state according to which is the  `justify-content: center` either added to the container or not.

I've tried to go through the docs to find an option (or combination of some) that would allow for this behaviour but not found any. Probably I miss sth, though. I think it is a very basic use case to support. 

The codesanbox provided is the very same as the generated one (React + horizontal + left to right + 20% slide width + 10px gap + align: center) , the only thing changed is the number of slides.

Any tips/helps is much appreciated 🙏  and thanks for developing and maintaining this library, it is awesome 💪 

### If applicable, which variants of Embla Carousel are relevant to this question?

- [ ] embla-carousel **(Core)**
- [X] embla-carousel-react
- [ ] embla-carousel-vue
- [ ] embla-carousel-svelte
- [ ] embla-carousel-autoplay
- [ ] embla-carousel-auto-scroll
- [ ] embla-carousel-solid
- [ ] embla-carousel-auto-height
- [ ] embla-carousel-class-names
- [ ] embla-carousel-fade
- [ ] embla-carousel-docs **(Documentation)**
- [ ] embla-carousel-docs **(Generator)**

### CodeSandbox example

https://codesandbox.io/p/sandbox/embla-carousel-generator-react-forked-pggc54

**Comment by davidjerleke** [comment]
@rudokotrla try [this CodeSandbox](https://codesandbox.io/p/sandbox/embla-carousel-generator-react-forked-skq4h6?file=%2Fsrc%2Fjs%2FEmblaCarousel.jsx%3A1%2C1).

**Reply by sarussss** [reply]
Hi @davidjerleke 
You can add a CodeSandbox for Vanilla JS?

**Reply by davidjerleke** [reply]
@sarussss sure, I can do that next week.

**Reply by sarussss** [reply]
@davidjerleke Thank you!
I tried a [CodeSandbox](https://codesandbox.io/p/sandbox/embla-carousel-generator-vanilla-forked-yydryr?file=%2Fsrc%2Fjs%2Findex.js%3A17%2C43).
When you have time, can you check and help me see if the code can be optimized?

**Reply by davidjerleke** [reply, marked as answer]
@sarussss I think your solution makes sense. Thanks for providing it!

- [Sandbox Vanilla](https://codesandbox.io/p/sandbox/embla-carousel-generator-vanilla-forked-xvyqt2) - Thanks @sarussss!
- [Sandbox React](https://codesandbox.io/p/sandbox/embla-carousel-generator-react-forked-skq4h6?file=%2Fsrc%2Fjs%2FEmblaCarousel.jsx%3A1%2C1)

**Comment by EdgarsJoja** [comment]
What seems to also do the trick is to use `safe` centering on container:
```css
justify-content: safe center;
```
See docs: https://developer.mozilla.org/en-US/docs/Web/CSS/justify-content#safe

tldr: this will only center slides when they do not overflow container.
